src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,117 @@ public SubstitutionMatrix(final byte[] matrix) {
         baseByCode['N'][(encodedMatrixBytes[4] >> 2) & 3] = 'G';
         baseByCode['N'][(encodedMatrixBytes[4]) & 3] = 'T';
 
-        for (final byte refBase : SUBSTITUTION_BASES_UPPER) {
+        for (final SubstitutionBase refBase : SubstitutionBase.values()) {
             for (byte code = 0; code < CODES_PER_BASE; code++) {
-                codeByBase[refBase][baseByCode[refBase][code]] = code;
+                codeByBase[refBase.getBase()][baseByCode[refBase.getBase()][code]] = code;
             }
         }
     }
 
+    /**
+     * Given a reference base and a read base, find the corresponding substitution code
+     * @param refBase reference base being substituted
+     * @param readBase read base to substitute for the reference base
+     * @return code to be used for this refBase/readBase pair
+     */
+    public byte code(final byte refBase, final byte readBase) {
+        if (Character.isLowerCase((char) refBase)) {
+            log.warn(String.format("CRAM: Attempt to generate a substitution code for a lower case reference base '%c'", (char) refBase));
```

Comment:
Is this log potentially going to penalize performance too much with some real-life datasets or you consider them as pathological? There is already and additional ```if``` evaluation penalty for everyone and for those that have a lower case base a string format and a log.

Reply:
This does seem weird.  If it's abnormal lets throw and we'll find out that it's happening.  If it's expected than logging on  evert read * base is a bad idea.  If it's expected than we can of course just do the substitution without comment.

Reply:
Yes, let throw. I'm hoping that at some point soon we can remove the lower case code entirely, since it really shouldn't happen. Having said that, I'm unsure why the original author went out of their way to make it work.

Reply:
Done.